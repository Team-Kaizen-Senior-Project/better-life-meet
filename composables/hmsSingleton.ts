import type { HmsInstance } from "~/types";
let hmsInstance: HmsInstance | null = null;

export function getHmsInstance(): HmsInstance {
    if (hmsInstance === null) {
      console.log("Creating a new instance of useHms");
      hmsInstance = useHms();
    } else {
      console.log("Reusing existing instance of useHms");
    }
    return hmsInstance;
  }
