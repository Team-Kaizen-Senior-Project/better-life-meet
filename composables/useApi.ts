import type {
	Attendee,
	AttendeeFields,
	Customer,
	CustomerFields,
	Meeting,
	MeetingFields,
	Numberic,
	Pod,
	PodFields,
} from '~/types'

// All the api requests, error and response handling should happen where they are being called.

type MeetingQueryParams =
	| { podId?: never; count?: number; cursor?: string }
	| { podId: Numberic; count?: number; cursor?: never }

export const useApi = () => {
	const createCustomer = async (customer: CustomerFields) => {
		const response = await $fetch<{ data: Customer }>('/api/customer', {
			method: 'POST',
			body: customer,
		})

		const newCustomer = response.data
		return newCustomer
	}

	const getCustomer = async (id: Numberic): Promise<Customer> => {
		const response = await $fetch<{ data: Customer }>(`/api/customer/${id}`)

		const customer = response.data
		return customer
	}

	const getCustomerByEmail = async (email: string): Promise<Customer> => {
		const response = await $fetch<{ data: Customer }>(`/api/customer?email=${email}`)

		const customer = response.data
		return customer
	}

	const updateCustomer = async (id: Numberic, customer: CustomerFields): Promise<Customer> => {
		const response = await $fetch<{ data: Customer }>(`/api/customer/${id}`, {
			method: 'PATCH',
			body: customer,
		})

		const updatedCustomer = response.data
		return updatedCustomer
	}

	const deleteCustomer = async (id: Numberic): Promise<unknown> => {
		const response = await $fetch(`/api/customer/${id}`, { method: 'DELETE' })

		return response
	}

	const createPod = async (pod: PodFields): Promise<Pod> => {
		const response = await $fetch<{ data: Pod }>('/api/pod', {
			method: 'POST',
			body: pod,
		})

		const newPod = response.data
		return newPod
	}

	const getPod = async (id: Numberic): Promise<Pod> => {
		const response = await $fetch<{ data: Pod }>(`/api/pod/${id}`)

		const pod = response.data
		return pod
	}

	const updatePod = async (id: Numberic, pod: PodFields): Promise<Pod> => {
		const response = await $fetch<{ data: Pod }>(`/api/pod/${id}`, {
			method: 'PATCH',
			body: pod,
		})

		const updatedPod = response.data
		return updatedPod
	}

	const deletePod = async (id: Numberic): Promise<unknown> => {
		const response = await $fetch(`/api/pod/${id}`, { method: 'DELETE' })

		return response
	}

	const createMeeting = async (meeting: MeetingFields): Promise<Meeting> => {
		console.log(meeting)
		const response = await $fetch<{ data: Meeting }>('/api/meeting', {
			method: 'POST',
			body: meeting,
		})

		const newMeeting = response.data
		return newMeeting
	}

	const getMeeting = async (id: Numberic): Promise<Meeting> => {
		const response = await $fetch<{ data: Meeting }>(`/api/meeting/${id}`)

		const meeting = response.data
		return meeting
	}

	const getMeetings = async (params?: MeetingQueryParams): Promise<Meeting[]> => {
		const response = await $fetch<{ data: { data: Meeting[] } }>(`/api/meeting`, { params })
		//console.log(response.data.data)
		// filter out meetings that have already ended
		// sort the meetings by start time so newest is first
		return response.data.data
			.filter((meeting) => typeof meeting.endTime != 'string' && new Date(meeting.endTime.isoString) > new Date())
			.sort((a, b) => a.startTime.isoString.localeCompare(b.startTime.isoString))
	}

	const updateMeeting = async (id: Numberic, meeting: MeetingFields): Promise<Meeting> => {
		const response = await $fetch<{ data: Meeting }>(`/api/meeting/${id}`, {
			method: 'PATCH',
			body: meeting,
		})

		const updatedMeeting = response.data
		return updatedMeeting
	}

	const deleteMeeting = async (id: Numberic): Promise<unknown> => {
		const response = await $fetch(`/api/meeting/${id}`, { method: 'DELETE' })

		return response
	}

	const createAttendee = async (attendee: AttendeeFields): Promise<Attendee> => {
		const response = await $fetch<{ data: Attendee }>('/api/attendee', {
			method: 'POST',
			body: attendee,
		})

		const newAttendee = response.data
		return newAttendee
	}

	const getAttendee = async (id: Numberic): Promise<Attendee> => {
		const response = await $fetch<{ data: Attendee }>(`/api/attendee/${id}`)

		const attendee = response.data
		return attendee
	}

	const updateAttendee = async (id: Numberic, attendee: AttendeeFields): Promise<Attendee> => {
		const response = await $fetch<{ data: Attendee }>(`/api/attendee/${id}`, {
			method: 'PATCH',
			body: attendee,
		})

		const updatedAttendee = response.data
		return updatedAttendee
	}

	const deleteAttendee = async (id: Numberic): Promise<unknown> => {
		const response = await $fetch(`/api/attendee/${id}`, { method: 'DELETE' })

		return response
	}

	return {
		// Customer
		createCustomer,
		getCustomer,
		getCustomerByEmail,
		updateCustomer,
		deleteCustomer,

		// Pod
		createPod,
		getPod,
		updatePod,
		deletePod,

		// MEETINGS
		getMeetings,
		deleteMeeting,
		createMeeting,
		getMeeting,
		updateMeeting,

		// Attendee
		createAttendee,
		getAttendee,
		updateAttendee,
		deleteAttendee,
	}
}
