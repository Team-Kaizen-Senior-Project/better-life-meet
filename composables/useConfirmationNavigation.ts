export default function useConfirmNavigation(message = 'Are you sure you want to leave this page?') {
	let isLeavingConfirmed = false // Flag to track confirmation status
	const confirmNavigation = (event) => {
		if (event.type === 'beforeunload') {
			console.log('tab closed triggered')

			// Triggered on tab close
			if (!isLeavingConfirmed) {
				if (!confirm(message)) {
					event.preventDefault()
					event.returnValue = ''
				} else {
					isLeavingConfirmed = true
				}
			}
		} else if (event.type === 'popstate') {
			// Triggered on back button
			console.log('back buttib triggered')

			if (!isLeavingConfirmed) {
				if (confirm(message)) {
					history.back()
				} else {
					event.preventDefault()
				}
			}
		}
	}

	const setupConfirmNavigation = () => {
		console.log('setting up event listeners')
		window.addEventListener('beforeunload', confirmNavigation, false)
		window.addEventListener('popstate', confirmNavigation, false)
	}

	const removeConfirmNavigation = () => {
		window.removeEventListener('beforeunload', confirmNavigation, false)
		window.removeEventListener('popstate', false)
	}

	return { setupConfirmNavigation, removeConfirmNavigation }
}
