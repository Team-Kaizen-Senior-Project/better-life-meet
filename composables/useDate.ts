import dayjs from 'dayjs'
import utc from 'dayjs/plugin/utc'
import timezone from 'dayjs/plugin/timezone'
import isBetween from 'dayjs/plugin/isBetween'
import duration from 'dayjs/plugin/duration'

export const useDate = () => {
	dayjs.extend(utc)
	dayjs.extend(timezone)
	dayjs.extend(isBetween)
	dayjs.extend(duration)

	const display = (date?: string) => {
		return dayjs(date).tz().format('MMM DD YYYY hh:mm a')
	}

	return {
		display,
		dayjs,
	}
}
