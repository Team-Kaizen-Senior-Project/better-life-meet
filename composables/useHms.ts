import { ref, onBeforeUnmount, onMounted, watch } from 'vue'
import {
	HMSReactiveStore,
	selectIsLocalAudioEnabled,
	selectIsLocalVideoEnabled,
	selectPeers,
	selectIsConnectedToRoom,
	selectVideoTrackByID,
	selectHMSMessages,
	selectBroadcastMessages,
} from '@100mslive/hms-video-store'
import type { HmsInstance, ChatMessage } from '~/types'
import { select } from '@nuxt/ui/dist/runtime/ui.config'

export function useHms(): HmsInstance {
	console.log('Initializing useHms instance')
	const hmsManager = new HMSReactiveStore()
	hmsManager.triggerOnSubscribe()
	const hmsStore = hmsManager.getStore()
	const hmsActions = hmsManager.getActions()

	const userName = ref('')
	const roomCode = ref('')
	const videoRefs = ref([])

	const isLocalAudioEnabled = ref(false)
	const isLocalVideoEnabled = ref(false)
	const isConnected = ref(false)
	const peers = ref([])
	const messages: Ref<ChatMessage[]> = ref([])
	const video = useVideoStore()

	watch(peers, (newPeers) => {
		// synchronize the video elements with the peers
		newPeers.forEach((peer, index) => {
			hmsActions.attachVideo(peer.videoTrack, videoRefs.value[index])
		})
	})

	onBeforeUnmount(() => {
		leaveRoom()
	})

	const joinRoom = async (roomCode: string, username: string) => {
		const authToken = await hmsActions.getAuthTokenByRoomCode({ roomCode })
		await hmsActions.join({
			userName: username,
			authToken,
			settings: {
				isAudioMuted: true,
				isVideoMuted: !video.cameraActive,
			},
		})
	}

	const leaveRoom = async () => {
		await hmsActions.leave()
	}

	const toggleAudio = async () => {
		console.log('Toggling audio')
		isLocalAudioEnabled.value = !isLocalAudioEnabled.value
		await hmsActions.setLocalAudioEnabled(isLocalAudioEnabled.value)
	}

	const toggleVideo = async () => {
		console.log('Toggling video')
		isLocalVideoEnabled.value = !isLocalVideoEnabled.value
		await hmsActions.setLocalVideoEnabled(isLocalVideoEnabled.value)
	}

	const sendBroadcastMessage = async (message: string) => {
		await hmsActions.sendBroadcastMessage(message)
	}

	hmsStore.subscribe((newMessages) => {
		if (newMessages.length > 0) {
			const mostRecentMessage = newMessages[newMessages.length - 1]
			messages.value.push({
				id: mostRecentMessage.id,
				content: mostRecentMessage.message,
			})
		}
	}, selectHMSMessages) //for all messages, send

	hmsStore.subscribe((val) => (isLocalAudioEnabled.value = val), selectIsLocalAudioEnabled)
	hmsStore.subscribe((val) => (isLocalVideoEnabled.value = val), selectIsLocalVideoEnabled)
	hmsStore.subscribe((val) => (isConnected.value = val), selectIsConnectedToRoom)
	hmsStore.subscribe((val) => (peers.value = val), selectPeers)

	return {
		userName,
		roomCode,
		videoRefs,
		isLocalAudioEnabled,
		isLocalVideoEnabled,
		isConnected,
		peers,
		messages,
		joinRoom,
		leaveRoom,
		toggleAudio,
		toggleVideo,
		sendBroadcastMessage,
	}
}
