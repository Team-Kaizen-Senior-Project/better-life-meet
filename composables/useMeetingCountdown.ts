interface Params {
	startTime?: MaybeRef<string | undefined>
	onMeetingStart?: () => any
}

export const useMeetingCountdown = (params: Params) => {
	const { display: displayDate, dayjs } = useDate()

	const countdown = ref<string | undefined>(undefined)
	const hasStarted = ref<boolean>(false)

	const interval = ref<NodeJS.Timeout | undefined>(undefined)

	const updateCountdown = () => {
		const unrefedTime = unref(params.startTime)

		if (!unrefedTime) {
			countdown.value = undefined
			console.log('no start time is provided')
			return
		}

		const startTime = dayjs(unrefedTime)
		const now = dayjs()

		if (now.isAfter(startTime)) {
			onTimerFinish()
			return
		}

		const duration = dayjs.duration(startTime.diff(now))
		countdown.value = `Starts in ${duration.days()}d ${duration.hours()}h ${duration.minutes()}m ${duration.seconds()}s`
		hasStarted.value = false
	}

	const runInterval = () => {
		clearInterval(interval.value)
		interval.value = setInterval(() => updateCountdown(), 1000)
		updateCountdown()
	}

	const stopInterval = () => {
		clearInterval(interval.value)
	}

	const onTimerFinish = () => {
		hasStarted.value = true
		params.onMeetingStart?.()
		stopInterval()
	}

	const renderText = computed(() => (!hasStarted.value ? countdown.value : 'Meeting Has Started'))

	onMounted(() => runInterval())
	onBeforeUnmount(() => stopInterval())

	return {
		hasStarted,
		countdown,
		renderText,
	}
}
