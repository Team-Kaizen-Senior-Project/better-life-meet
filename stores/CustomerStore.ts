import type { Customer, Pod } from '~/types'
interface State {
	customer?: Customer
}
export const useCustomerStore = defineStore('customer', () => {
	// TODO: This is unused
	const state = reactive<State>({
		customer: undefined,
	})

	const getMe = async (): Promise<Customer> => {
		const { getSession } = useAuth()
		const session = await getSession()
		const customer = session.user as Customer

		state.customer = customer

		return customer
	}

	return {
		getMe,
		state,
	}
})
