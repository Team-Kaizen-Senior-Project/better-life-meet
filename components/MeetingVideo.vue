<script setup>
	import { useHms } from '~/composables/useHms.ts'
	import { onMounted } from 'vue'

	const customerStore = useCustomerStore()
	const customer = customerStore.state.customer

	const {
		videoRefs,
		isLocalAudioEnabled,
		isLocalVideoEnabled,
		isConnected,
		peers,
		joinRoom,
		leaveRoom,
		toggleAudio,
		toggleVideo,
		sendChatMessage,
	} = getHmsInstance()

	const roomCode = 'qas-jmqz-old'
	let userName = ''

	onMounted(async () => {
		if (customer?.firstName && customer?.lastName) {
			userName = customer.firstName + ' ' + customer.lastName
			await joinRoom(roomCode, userName)
			await sendChatMessage('hello')
		}
	})

	window.addEventListener('beforeunload', () => {
		if (isConnected.value) {
			leaveRoom()
		}
	})
</script>

<template>
	<div class="container mx-auto mb-8 mt-8">
		<div v-if="isConnected" class="conference-section">
			<div class="grid grid-cols-1 gap-4 sm:grid-cols-2 lg:grid-cols-3">
				<div v-for="peer in peers" :key="peer.id" class="peer-tile w-full rounded-lg bg-zinc-500 p-4 sm:w-auto">
					<video ref="videoRefs" class="peer-video mx-auto h-auto w-full rounded-lg" autoplay muted playsinline></video>
					<div class="peer-name mt-2 text-center font-bold text-black">{{ peer.name }}</div>
				</div>
			</div>
		</div>
	</div>
</template>
