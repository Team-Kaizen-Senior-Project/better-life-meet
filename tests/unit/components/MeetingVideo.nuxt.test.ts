import { describe, it, expect, vi } from 'vitest'
import { mountSuspended } from '@nuxt/test-utils/runtime'
import MeetingVideo from '@/components/MeetingVideo.vue'
import { nextTick } from 'vue'
import { waitFor } from '@testing-library/vue'
import { useCustomerStore } from "~/stores/CustomerStore"
import { getHmsInstance } from "~/composables/useHms"

vi.mock('~/stores/CustomerStore', () => ({
    useCustomerStore: vi.fn(() => ({
        state: {
            customer: {
                firstName: 'John',
                lastName: 'Doe'
            }
        }
    }))
}))

vi.mock('~/composables/useHms', () => ({
    useHms: vi.fn(() => ({
        videoRefs: vi.fn(),
        isLocalAudioEnabled: true,
        isLocalVideoEnabled: true,
        isConnected: true,
        peers: [
            { id: '1', name: 'peer 1' },
            { id: '2', name: 'peer 2' },
            { id: '3', name: 'peer 3' }
        ],
        joinRoom: vi.fn(),
        leaveRoom: vi.fn(),
        toggleAudio: vi.fn(),
        toggleVideo: vi.fn(),
        dominantSpeaker: { id: '1', name: 'peer 1' },
        peersWithAudioStatus: [
            { peer: { id: '1', name: 'peer 1' }, isAudioEnabled: true },
            { peer: { id: '2', name: 'peer 2' }, isAudioEnabled: true },
            { peer: { id: '3', name: 'peer 3' }, isAudioEnabled: true }
        ]
    }))
}))

describe('MeetingVideo', () => {
    it('renders correctly', async () => {
        let wrapper
        try {
            wrapper = await mountSuspended(MeetingVideo, {
                global: {
                    plugins: []
                }
            })
            expect(isMounted(wrapper)).toBe(true)
        } catch (error) {
            console.error('Error mounting component:', error)
        }
    })

    it('Video renders', async () => {
        const wrapper = await mountSuspended(MeetingVideo, {
            global: {
                plugins: []
            }
        })
        await nextTick()
        expect(wrapper.find('[data-testid="video"]').exists()).toBe(true)
    });

    it ('Checks peers`', async () => {
        const wrapper = await mountSuspended(MeetingVideo, {
            global: {
                plugins: []
            }
        })
        await nextTick()
        const peerNames = wrapper.findAll('.peer-name')
        expect(peerNames.length).toBe(3)
    })

    it ('creates video elements for each peer', async () => {
        const wrapper = await mountSuspended(MeetingVideo, {
            global: {
                plugins: []
            }
        })
        await nextTick()
        const videoElements = wrapper.findAll('video')
        expect(videoElements.length).toBe(3)
    })

    it ('uses correct user name when joining room', async () => {
        const wrapper = await mountSuspended(MeetingVideo, {
            global: {
                plugins: []
            }
        })
        await nextTick()
        const c = useCustomerStore()
        const expectedUserName = c.state.customer?.firstName + ' ' + c.state.customer?.lastName
        expect(expectedUserName).toBe('John Doe')
    })
});