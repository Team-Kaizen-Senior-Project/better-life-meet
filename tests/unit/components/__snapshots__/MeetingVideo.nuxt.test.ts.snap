// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MeetingVideo > renders correctly 1`] = `
"<div class="container mx-auto mb-8 mt-8">
  <!--v-if-->
</div>"
`;
